quantum/process_keycode/process_swap_hands.c

```diff
@@ -0,0 +1,155 @@
+/* Copyright 2021 QMK
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "process_swap_hands.h"
+#include "quantum_keycodes.h"
+#include "action_tapping.h"
+#include "matrix.h"
+
+extern const keypos_t PROGMEM hand_swap_config[MATRIX_ROWS][MATRIX_COLS];
+
+static matrix_row_t swap_state[MATRIX_ROWS];
+
+static bool     swap_hands_enabled = false;
+static bool     oneshot_enabled    = false;
+static uint16_t oneshot_expiry     = 0;
+
+#if defined(ONESHOT_TIMEOUT) && (ONESHOT_TIMEOUT > 0)
+static inline bool has_oneshot_swaphands_timed_out(void) {
+    return (oneshot_expiry && timer_expired(timer_read(), oneshot_expiry));
+}
+#endif
+
+static inline void swap_hands_task(void) {
+#if defined(ONESHOT_TIMEOUT) && (ONESHOT_TIMEOUT > 0)
+    if (has_oneshot_swaphands_timed_out()) {
+        clear_oneshot_swaphands();
+    }
+#endif
+}
+
+void clear_oneshot_swaphands(void) {
+    oneshot_expiry     = 0;
+    swap_hands_enabled = false;
+    oneshot_enabled    = false;
+}
+
+void preprocess_swap_hands(keyrecord_t *record) {
+    // TODO: implement a "real" swap_hands_task called from matrix_scan_quantum?
+    swap_hands_task();
+
+    keypos_t     pos     = record->event.key;
+    matrix_row_t col_bit = MATRIX_ROW_SHIFTER << pos.col;
+    bool         do_swap = record->event.pressed ? swap_hands_enabled : swap_state[pos.row] & (col_bit);
+
+    if (do_swap) {
+        record->event.key.row = pgm_read_byte(&hand_swap_config[pos.row][pos.col].row);
```

Comment:
Is it possible to make this more granular than before so it behaves different on different layers as it is now its ok for characters but there are some corner cases or am I thinking of this feature wrongly because I mirror the layer. 
I am not used to utilize a single hand so I don't know if it's the correct way but my theory was on a QWERTY  keyboard my forefinger rest on "J" on my right side so it should be the same finger on the left side.

Some examples I thought of while testing because I have only had one side of keyboard.

Arrow keys or Mouse cursor keys
Disabled swap hands
        ____
       |  /\ |
       |  ||  |
 ____ ____ ___
|<-  |  ||  | -> |
|____| \/  |____|

Enabled swap hands
        ____
       |  /\ |
       |  ||  |
 ____ ____ ___
|->  |  ||  | <- |
|____| \/  |____|

Numpad example

Disabled swap hands
/ *  -
7 8 9 +
4 5 6 E
1 2 3 N
0    .  T

Disabled swap hands
-  *  /
+ 9 8 7
E  6 5 4
N 3 2 1
T  .  0

Reply:
Not really.  It remaps row and column positions, and then allows those to be called/mapped.  So there is no way to do that.

Also, this isn't really the correct place to discuss this.

Reply:
:+1

Reply:
Given there are existing issues with getting the code migrated over, I'm not going to consider any extensions to scope at this point.